shell/browser/win/dark_mode.cc

```diff
@@ -1,178 +1,51 @@
-  }
-
-  // load "SetWindowCompositionAttribute", used in RefreshTitleBarThemeColor()
-  _SetWindowCompositionAttribute =
-      reinterpret_cast<decltype(_SetWindowCompositionAttribute)>(
-          base::win::GetUser32FunctionPointer("SetWindowCompositionAttribute"));
-  if (_SetWindowCompositionAttribute == nullptr) {
-    return;
-  }
-
-  // load the dark mode functions from uxtheme.dll
-  // * RefreshImmersiveColorPolicyState()
-  // * ShouldAppsUseDarkMode()
-  // * AllowDarkModeForApp()
-  // * SetPreferredAppMode()
-  // * AllowDarkModeForApp() (build < 18362)
-  // * SetPreferredAppMode() (build >= 18362)
-
-  base::NativeLibrary uxtheme =
-      base::PinSystemLibrary(FILE_PATH_LITERAL("uxtheme.dll"));
-  if (!uxtheme) {
-    return;
-  }
-  auto ux_pei = base::win::PEImage(uxtheme);
-  auto get_ux_proc_from_ordinal = [&ux_pei](int ordinal, auto* setme) {
-    FARPROC proc = ux_pei.GetProcAddress(reinterpret_cast<LPCSTR>(ordinal));
-    *setme = reinterpret_cast<decltype(*setme)>(proc);
-  };
-
-  // ordinal 104
-  using fnRefreshImmersiveColorPolicyState = VOID(WINAPI*)();
-  fnRefreshImmersiveColorPolicyState _RefreshImmersiveColorPolicyState = {};
-  get_ux_proc_from_ordinal(104, &_RefreshImmersiveColorPolicyState);
-
-  // ordinal 132
-  using fnShouldAppsUseDarkMode = BOOL(WINAPI*)();
-  fnShouldAppsUseDarkMode _ShouldAppsUseDarkMode = {};
-  get_ux_proc_from_ordinal(132, &_ShouldAppsUseDarkMode);
-
-  // ordinal 135, in 1809
-  using fnAllowDarkModeForApp = BOOL(WINAPI*)(BOOL allow);
-  fnAllowDarkModeForApp _AllowDarkModeForApp = {};
-
-  // ordinal 135, in 1903
-  typedef PreferredAppMode(WINAPI *
-                           fnSetPreferredAppMode)(PreferredAppMode appMode);
-  fnSetPreferredAppMode _SetPreferredAppMode = {};
-
-  if (g_buildNumber < 18362) {
-    get_ux_proc_from_ordinal(135, &_AllowDarkModeForApp);
-  } else {
-    get_ux_proc_from_ordinal(135, &_SetPreferredAppMode);
-  }
-
-  // dark mode is supported iff we found the functions
-  g_darkModeSupported = _RefreshImmersiveColorPolicyState &&
-                        _ShouldAppsUseDarkMode &&
-                        (_AllowDarkModeForApp || _SetPreferredAppMode);
-  if (!g_darkModeSupported) {
-    return;
+  const BOOL isDarkMode = dark;
+  if (FAILED(DwmSetWindowAttribute(hWnd, DWMWA_USE_IMMERSIVE_DARK_MODE,
+                                   &isDarkMode, sizeof(isDarkMode)))) {
+    result =
+        DwmSetWindowAttribute(hWnd, DWMWA_USE_IMMERSIVE_DARK_MODE_BEFORE_20H1,
+                              &isDarkMode, sizeof(isDarkMode));
+    if (FAILED(result))
+      return result;
   }
 
-  // initial setup: allow dark mode to be used
-  if (_AllowDarkModeForApp) {
-    _AllowDarkModeForApp(true);
-  } else if (_SetPreferredAppMode) {
-    _SetPreferredAppMode(AllowDark);
-  }
-  _RefreshImmersiveColorPolicyState();
-
-  // check to see if dark mode is currently enabled
-  g_darkModeEnabled = _ShouldAppsUseDarkMode() && !IsHighContrast();
+  return S_OK;
```

Comment:
```suggestion
  // Toggle the nonclient area active state to force the titlebar to redraw
  SendMessage(hWnd, WM_NCACTIVATE, FALSE, 0);
  SendMessage(hWnd, WM_NCACTIVATE, TRUE, 0);

  return S_OK;
```

Reply:
@deermichel, in my testing this fixes the issue where the titlebar doesn't update. 🙂

It's a bit hackish, but it's functional. My gut tells me there's a more elegant way to accomplish this, and it also tells me there's a rabbit hole with Chromium since it handles a lot of the window messages. If there's a root cause anywhere, I'd wager it might be found in the Chromium codebase, in [`ui/views/win/hwnd_message_handler.cc`](https://source.chromium.org/chromium/chromium/src/+/main:ui/views/win/hwnd_message_handler.cc). Only way to know for sure would be some time testing the theme change out in a simple "Hello World" Windows app and see if there's a difference in behavior, and go from there. Unfortunately I don't have the time to run that down at the moment, but this works, and I didn't see any negative effects from it.

Reply:
Can you try it on Win11 please? It seems to me that this has been fixed in Windows (probably that's why the property was not officially documented before).

Reply:
Tried it on Win11, you're right, doesn't look like that toggling is necessary. We can wrap it with an if guard and only do it on Win10.

Reply:
allright, i don't have a win10 instance around to verify - but i'm ok keeping that code for win10